ui/pages/destinations/grantAccessContent.js

```diff
@@ -0,0 +1,151 @@
+import useSWR, { useSWRConfig } from 'swr'
+import { useState } from 'react'
+import styled from 'styled-components'
+
+import { validateEmail } from '../../lib/email'
+
+import InputDropdown from '../../components/input'
+import ErrorMessage from '../../components/error-message'
+
+const GrantList = styled.section`
+  max-height: 20rem;
+  overflow: auto;
+  width: 95%;
+  padding: 0 1.25rem;
+`
+
+const GrantListItem = styled.div`
+  display: flex;
+  flex-direction: row;
+  justify-content: space-between;
+  align-items: center;
+`
+
+const Grant = ({ id }) => {
+  const { data: user } = useSWR(`/v1/identities/${id.replace('i:', '')}`, { fallbackData: { name: '' } })
+
+  return (
+    <p>{user.name}</p>
+  )
+}
+
+export default ({ id }) => {
+  const { data: destination } = useSWR(`/v1/destinations/${id}`)
+  const { data: list } = useSWR(() => `/v1/grants?resource=${destination.name}`)
+  const { mutate } = useSWRConfig()
+
+  const [email, setEmail] = useState('')
+  const [error, setError] = useState('')
+  const [role, setRole] = useState('view')
+
+  const options = ['view', 'edit', 'admin', 'remove']
+
+  const grantPrivilege = (id, privilege = role) => {
+    fetch('/v1/grants', {
+      method: 'POST',
+      body: JSON.stringify({ subject: id, resource: destination.name, privilege })
+    })
+      .then((response) => response.json())
+      .then(() => mutate(`/v1/grants?resource=${destination.name}`))
+      .finally(() => setEmail(''))
+  }
+
+  const handleInputChang = (value) => {
+    setEmail(value)
+    setError('')
+  }
+
+  const handleKeyDownEvent = (key) => {
+    if (key === 'Enter' && email.length > 0) {
+      handleShareGrant()
+    }
+  }
+
+  const handleShareGrant = () => {
+    if (validateEmail(email)) {
+      setError('')
+      fetch(`/v1/identities?name=${email}`)
+        .then((response) => response.json())
+        .then((data) => {
+          if (data.length === 0) {
+            fetch('/v1/identities', {
+              method: 'POST',
+              body: JSON.stringify({ name: email, kind: 'user' })
+            })
+              .then((response) => response.json())
+              .then((user) => grantPrivilege('i:' + user.id))
+              .finally(() => setEmail(''))
+          } else {
+            grantPrivilege(data[0].id)
+          }
+        })
+        .catch((error) => console.error(error))
+    } else {
+      setError('Invalid email')
+    }
+  }
+
+  const handleUpdateGrant = (privilege, grantId, userId) => {
+    fetch(`/v1/grants/${grantId}`, { method: 'DELETE' })
+      .then(() => {
+        if (privilege === 'remove') {
+          mutate(`/v1/grants?resource=${destination.name}`)
```

Comment:
Better to put the fetch in `mutate` so we can have an [optimistic update](https://swr.vercel.app/docs/mutation#optimistic-updates) (instant to the user):